### PR TITLE
Add optional printing for effects

### PR DESCRIPTION
**Pull request by ianatol**
This adds optional printing for effects by callsite. Press `e` to turn them on or off!

CC: @Keno @aviatesk

### CODE REVIEW

File: src/reflection.jl [hunk 1]
```diff
@@ -130,20 +132,21 @@ function process_info(interp, @nospecialize(info), argtypes::ArgTypes, @nospecia
         infos = process_recursive(info.call)
         @assert length(infos) == length(info.results)
         return mapany(enumerate(info.results)) do (i, result)
-            if isnothing(result)
+            if isnothing(result) || isa(result, Bool)
```
Comment (simeonschaub) [inline src/reflection.jl:135]: I'm curious, what does a `Bool` here mean?

File: src/Cthulhu.jl [hunk 2]
```diff
@@ -294,10 +303,10 @@ function _descend(term::AbstractTerminal, interp::CthulhuInterpreter, mi::Method
             # This was inferred to a pure constant - we have no code to show,
             # but make something up that looks plausible.
             if display_CI
-                println(iostream)
-                println(iostream, "│ ─ $(string(Callsite(-1, MICallInfo(mi, interp.opt[mi].rettype), :invoke)))")
-                println(iostream, "│  return ", Const(interp.opt[mi].rettype_const))
-                println(iostream)
+                println(term.out_stream::IO)
```
Comment (simeonschaub) [inline src/Cthulhu.jl:306]: Why did you change all these?

File: src/Cthulhu.jl [hunk 3]
```diff
@@ -344,11 +354,11 @@ function _descend(term::AbstractTerminal, interp::CthulhuInterpreter, mi::Method
 
             if display_CI
                 _remarks = remarks ? get(interp.remarks, mi, nothing) : nothing
-                printstyled(IOContext(iostream, :limit=>true), mi.def, '\n'; bold=true)
+                printstyled(IOContext(term.out_stream::IO, :limit=>true), '[', effects ,']', mi.def, '\n'; bold=true)
                 if debuginfo == DInfo.compact
                     str = let debuginfo=debuginfo, src=src, codeinf=codeinf, rt=rt, mi=mi,
                               iswarn=iswarn, hide_type_stable=hide_type_stable,
-                              remarks=_remarks, inline_cost=inline_cost, type_annotations=type_annotations
+                              remarks=_remarks,  inline_cost=inline_cost, type_annotations=type_annotations
```
Comment (simeonschaub) [inline src/Cthulhu.jl:361]: extra space

**Comment by ianatol** [comment]
Missing a couple constructors, fixing now

File: src/reflection.jl (hunk 1 above)
Comment (ianatol) [reply src/reflection.jl:135]: Not 100% sure, I left it in while rebasing from https://github.com/JuliaDebug/Cthulhu.jl/pull/266/files#diff-11ccfcf41873e6a1cf01d00ed11aee2a59b7c1c1634cfa80c84b8049c6c34564R137-R143

**Comment by simeonschaub** [comment]
Hope it's ok I pushed a couple fixes here for errors I encountered when trying this out.

**Comment by ianatol** [comment]
Of course! Thank you

File: src/reflection.jl (hunk 1 above)
Comment (Keno) [reply src/reflection.jl:135]: Bool was dropped from the final version of the base change, so should be dropped here also.

File: src/Cthulhu.jl (hunk 2 above)
Comment (ianatol) [reply src/Cthulhu.jl]: Hmm, maybe a rebase error, I changed them back though, thanks

**Comment by ianatol** [comment]
I think this is ready

File: src/ui.jl [hunk 4]
```diff
@@ -96,6 +97,9 @@ function TerminalMenus.keypress(m::CthulhuMenu, key::UInt32)
     elseif key == UInt32('r')
         m.toggle = :remarks
         return true
+    elseif key == UInt32('f')
```
Comment (aviatesk) [inline src/ui.jl:100]: I think it would be more reasonable to change this keybinding to `e` and remove it from the `:edit` toggle. We had a similar discussion when we introduced `r`emark toggle and removed the `r`-binding from the Revise toggle.

Comment (ianatol) [reply src/ui.jl]: Sure, done!

File: src/reflection.jl [hunk 5]
```diff
@@ -134,16 +136,17 @@ function process_info(interp, @nospecialize(info), argtypes::ArgTypes, @nospecia
...
             elseif (@static isdefined(Compiler, :ConstResult) && true) && isa(result, Compiler.ConstResult)
                 linfo = result.mi
-                ConstEvalCallInfo(MICallInfo(linfo, rt), argtypes)
+                mici = MICallInfo(linfo, rt, haskey(interp.unopt, linfo) ? interp.unopt[linfo].effects : Effects())
+                ConstEvalCallInfo(mici, argtypes)
             else
                 @assert isa(result, Compiler.InferenceResult)
                 linfo = result.linfo
-                mici = MICallInfo(linfo, rt)
+                mici = MICallInfo(linfo, rt, result.ipo_effects)
```
Comment (simeonschaub) [inline src/reflection.jl:144]: This needs to be conditional on the Julia version

Comment (ianatol) [reply src/reflection.jl:144]: 👍  I'll refactor a bit to make this backwards compatible. Thanks for the review(s)!

**Comment by ianatol** [comment]
Ok, this should pass all CI except for the Integration test (I think the version of JET in that test needs to be bumped to work with 1.8)

**Review by vchuravy — APPROVED** [review]
